src/common/imagwebp.cpp

```diff
@@ -0,0 +1,231 @@
+
+// For compilers that support precompilation, includes "wx.h".
+#include "wx/wxprec.h"
+
+#if wxUSE_IMAGE && wxUSE_LIBWEBP
+
+#include "wx/imagwebp.h"
+#include "webp/demux.h"
+#include "webp/decode.h"
+#include "webp/encode.h"
+#include <vector>
+
+#ifndef WX_PRECOMP
+    #include "wx/intl.h"
+    #include "wx/log.h"
+#endif
+
+//-----------------------------------------------------------------------------
+// wxWEBPHandler
+//-----------------------------------------------------------------------------
+
+wxIMPLEMENT_DYNAMIC_CLASS(wxWEBPHandler, wxImageHandler);
+
+#if wxUSE_STREAMS
+
+#include <wx/mstream.h>
+#include <memory>
+#include <functional>
+
+typedef std::unique_ptr<WebPDemuxer, std::function<void(WebPDemuxer*)>> WebPDemuxerPtr;
+
+bool DecodeWebPDataIntoImage(wxImage *image, WebPData *webp_data, bool verbose) {
+    WebPBitstreamFeatures features;
+    VP8StatusCode status = WebPGetFeatures(webp_data->bytes, webp_data->size, &features);
+    if (status != VP8_STATUS_OK)
+    {
+        if (verbose)
+        {
+           wxLogError("WebP: GetFeatures not OK.");
+        }
+        return false;
+    }
+    image->Create(features.width, features.height, false); // this allocates memory
+    if (features.has_alpha)
+    {
+        // image has alpha channel. needs to be decoded, then re-ordered.
+        uint8_t * rgba = WebPDecodeRGBA(webp_data->bytes, webp_data->size, &features.width, &features.height);
+        if (nullptr == rgba)
+        {
+            if (verbose)
+            {
+               wxLogError("WebP: WebPDecodeRGBA failed.");
+            }
+            return false;
+        }
+        image->InitAlpha();
+        unsigned char * rgb = image->GetData();
+        unsigned char * alpha = image->GetAlpha();
+        size_t rgba_index = 0, rgb_index = 0, alpha_index = 0;
+        for (int pixel_counter = 0; pixel_counter < image->GetWidth() * image->GetHeight(); pixel_counter++)
+        {
+            rgb[rgb_index++] = rgba[rgba_index++]; // R
+            rgb[rgb_index++] = rgba[rgba_index++]; // G
+            rgb[rgb_index++] = rgba[rgba_index++]; // B
+            alpha[alpha_index++] = rgba[rgba_index++]; // A
+        }
+        WebPFree(rgba);
+    }
+    else
+    {
+        // image has no alpha channel. decode into target buffer directly.
+        int buffer_size = image->GetWidth() * image->GetHeight() * 3;
+        int stride = image->GetWidth() * 3;
+        uint8_t * output_buffer = WebPDecodeRGBInto(webp_data->bytes, webp_data->size, image->GetData(), buffer_size, stride);
+        if (nullptr == output_buffer)
+        {
+            if (verbose)
+            {
+               wxLogError("WebP: WebPDecodeRGBInto failed.");
+            }
+            return false;
+        }
+    }
+    image->SetMask(false); // all examples do this, so I do so as well
+    return true;
+}
+
+bool DecodeWebPFrameIntoImage(wxImage *image, int index, WebPDemuxerPtr & demuxer, bool verbose)
```

Comment:
It's unusual and almost always unnecessary to pass (smart) pointers by reference, why not just pass `WebPDemuxer&` directly here?